### ISSUE TITLE
Is this a bug in new CopyQ version?

### ISSUE THREAD

**Issue by Searcher1231** [hluk/CopyQ#2084]
Hi, dear.
My problem is exactly the same as mentioned in the following Post:

https://github.com/hluk/CopyQ/issues/1100

but the code that was given at it does not work for me. I don't know if it is a bug in new version of CopyQ or an incompatibility between CopyQ and the new version of my browser!?

When I press F4, the following formats are listed for my images:

text/plain
text/html

Thanks for any help
Best Regards

**Comment by hluk** [comment]
Can you be more specific? Do you mean images are not shown in CopyQ?

Can you check installed plugins in Items tab in Preferences dialog? It should list "Images".

**Comment by Searcher1231** [comment]
Yes, exactly. Images (albeit just Images that I have copied them with Firefox or Chrome's "Copy Image" option) have not any thumbnail and I can not paste them as an image in other programs (e.g. PhotoShop, PaintShop, ...). Other images have not any problem and are copied by CopyQ correctly.

Yes I have "Images" under CopyQ -> Preferences -> Items. 
![2022-09-12_18-14-06](https://user-images.githubusercontent.com/61396799/189670518-e12a1f86-a0ec-4726-9f9d-297f18f748c3.jpg)

**Comment by hluk** [comment]
Can you check clipboard if there are really any image formats after you copy them from web browser? (You can check current clipboard format with Ctrl+Shift+C from CopyQ.)

**Comment by Searcher1231** [comment]
> Can you check clipboard if there are really any image formats after you copy them from web browser? (You can check current clipboard format with Ctrl+Shift+C from CopyQ.)

Well, I'm sure there is an image formats in clipboard after copying them from web browser, because I can paste the last copied image from clipboard to any graphic program without any problem, however, CopyQ can not show the last copied image too!

BTW, I used Ctrl+Shift+C from CopyQ and there is a snapshot for you:

![2022-09-12_19-09-10](https://user-images.githubusercontent.com/61396799/189684128-c9b571a5-7658-4a28-861b-9ecb96172faa.jpg)

**Comment by hluk** [comment]
> Well, I'm sure there is an image formats in clipboard after copying them from web browser, because I can paste the last copied image from clipboard to any graphic program without any problem, however, CopyQ can not show the last copied image too!

It does not look like there is any image data CopyQ can use in the clipboard. Some applications can get the remote image from the HTML formatted data (or maybe from the uri-list). But CopyQ by default does not make any network requests.

BTW, what is the web browser? I tested it in Firefox 104.0.1 on Fedora 36 and it provides some standard image data in clipboard so CopyQ saves them as images correctly.

**Comment by Searcher1231** [comment]
> It does not look like there is any image data CopyQ can use in the clipboard. Some applications can get the remote image from the HTML formatted data (or maybe from the uri-list). But CopyQ by default does not make any network requests.

> BTW, what is the web browser? I tested it in Firefox 104.0.1 on Fedora 36 and it provides some standard image data in clipboard so CopyQ saves them as images correctly.

**I tested by FireFox_v91.13.0esr_64-bit, Chrome_v105.0.5195.102 (Official Build 64-bit) and Edge_v105.0.1343.33 (Official build 64-bit). My main Browser is FireFox. For example at the following link, no one of pictures can be copied by "Copy Image" **
https://www.jaaar.com/kiosk#rows/all/

**Comment by Searcher1231** [comment]
> I tested it in Firefox 104.0.1 on Fedora 36 and it provides some standard image data in clipboard so CopyQ saves them as images correctly.

**After your message, I thought maybe that is a problem with EAR version of Firefox, so I tested with the normal version (Firefox_v104.0.1). But the problem was not solved. Everything was exactly the same as previous.**

**Comment by hluk** [comment]
I think the solution would be similar to this: https://github.com/hluk/CopyQ/issues/1591#issuecomment-983280874

**Comment by Searcher1231** [comment]
> I think the solution would be similar to this: [#1591 (comment)](https://github.com/hluk/CopyQ/issues/1591#issuecomment-983280874)

**The problem still remains**

**Comment by latipun7** [comment]
I'm using Brave browser on **Linux**, this is not only Windows now, imo.
Here's is my reproduction step:
- right click image and "Copy image"
- see the content of clipboard using <kbd>Ctrl+Shift+C</kbd>
- the clipboard really showing the image
![image](https://user-images.githubusercontent.com/20012970/193096573-d2fe91a3-32ac-45b8-8054-0462e543d943.png)
- see the content of CopyQ of that image, no image
![image](https://user-images.githubusercontent.com/20012970/193097670-e5c28ebe-59ba-4ed3-b60f-d62cb342a7bb.png)

**Comment by hluk** [comment]
@latipun7 That looks like a different issue. Did you enable "Images" plugin in Items configuration tab in Preferences? Otherwise CopyQ will not store images. It is enabled by default unless plugins are not installed at all (sometimes there is a separate package `copyq-plugins`).

**Comment by latipun7** [comment]
@hluk , yes, it is enabled. I'm able to see image copied to clipboard from flameshot for example. But not from browser "Copy Image". Did I miss some requirement or something? I just installed Linux and copyq (fresh install).
![image](https://user-images.githubusercontent.com/20012970/193294418-e01cc032-913b-4c28-8784-1104510cac62.png)

When I click "Copy Image" in browser, my clipboard really did has image, so I can paste here in github comment for example. But if I copy from copyQ, when I paste here, it only the link/plain text.
Here's my clipboard in CopyQ looks like:
![image](https://user-images.githubusercontent.com/20012970/193298029-500fe304-8c1c-406c-ad64-3f2e4c71b351.png)

**Comment by hluk** [comment]
@latipun7 The image is probably not stored because of an automatic command. But that is a different issue, because you have a usable image format in clipboard. In the original issues there seems to be no image data in the system clipboard.

**Comment by latipun7** [comment]
oh, what should I do, @hluk ? Should I open new issue? Because I don't know what is blocking this image to be saved in CopyQ. I only has 1 custom auto command.
![image](https://user-images.githubusercontent.com/20012970/193330956-c01ec624-2a69-4a4d-bd9b-24c98640c53b.png)

**Comment by hluk** [comment]
@latipun7 Please open a new issue.

**Comment by NA0341** [comment]
> @hluk The image is probably not stored because of an automatic command. But that is a different issue, because you have a usable image format in clipboard. In the original issues there seems to be no image data in the system clipboard.

---

This issue is **not** related to:
+ Any custom settings or commands (config).
+ Specific operating sytems.

**But it is definitely related to Chromium based WebBrowsers!** (See Palemoon Screenshot below for reference.)

I've had this issue for quite a while now: in Chromium, Brave¹, Vivaldi, Ungoogled Chromium, and some more.
I added `.bak` to CopyQ's folders to start it freshly² and make some samples that can be seen below.

Note: Using Vivaldi's screenshot feature with »Copy to Clipboard« option works as intended (even with custom automatic commands).

---

**System Infomation:**

**Desktop:** `GNOME 44`
**OS:** `Manjaro Linux (rolling release)`
**Kernel:** `6.1.31-2-MANJARO #1 SMP PREEMPT_DYNAMIC`
**Laptop:** `Dell Latitude E5470`

---

¹ Don't take this one for granted.
² To `~/.local/share/copyq` and `~/.config/copyq`

**Copied Image from Vivaldi:**

![Image copied from Vivaldi (Screenshot from 2023-07-03 18-43-31)](https://github.com/hluk/CopyQ/assets/26059513/33262404-ebb6-44e8-acb3-60f98a924030)


**Same Image copied from Palemoon:**

![Image copied from Palemoon (Screenshot from 2023-07-03 18-51-26)](https://github.com/hluk/CopyQ/assets/26059513/e471ceb1-83ee-41de-bff3-cb043635722c)


**Image copied from the Site in [This Comment](https://github.com/hluk/CopyQ/issues/2084#issuecomment-1245033679) using Vivaldi:**

![Image copied from Vivaldi (Screenshot from 2023-07-03 18-35-14)](https://github.com/hluk/CopyQ/assets/26059513/458b4297-c48a-40f4-8c21-16544b36501e)

**Comment by jadedgnome** [comment]
i'm experiencing the same thing. preview of screenshot images show up in copyq items just fine but the moment I "copy image" from chrome on windows. I get a small white box instead. 

I used copyq-v3.9.3 for a long time without this sort of issue but I just recently noticed(using v6.3.2) that this white square box appears instead if the image. the only time i use to see that white square box on version 3.9.3 is when I did 'copy image' on a really high resolution image. 

![image](https://github.com/hluk/CopyQ/assets/2940805/26331ee5-f1b0-4471-948f-dc2d6cde9609)

the tabs I have with images stored in clipboard saved with 3.9.3 still appear normally in 6.3.2

**Comment by iG8R** [comment]
Have the same issue:(

**Version, OS and Environment**
 - CopyQ 7.1.0
 - OS: Microsoft Windows [Version 10.0.19045.2673] 64bit
 - Browser: Firefox 119.0 64bit
 - Browser: Chrome 118.0.5993.90 (Official Build) (64-bit)
 - Browser: Edge 118.0.2088.69 (Official build) (64-bit)
 
[CopyQ-Browsers-Images-issue-Video_2023-10-27_072615.webm](https://github.com/hluk/CopyQ/assets/11407417/62cb1172-429b-4338-9fe9-18dfb22528fb)

**Comment by hluk** [comment]
I don't think I can do anything about this since the clipboard (and basically Qt framework that CopyQ uses) does not seem to provide the image data.

For me on Linux in Firefox 118.0.2, right clicking on an image on a web page and selecting Copy Image, works well. There is a variety of image formats in the clipboard:

```
❯ copyq clipboard \?
text/html
text/_moz_htmlinfo
text/_moz_htmlcontext
image/png
image/jpeg
image/bmp
image/x-bmp
image/x-MS-bmp
image/x-icon
image/x-ico
image/x-win-bitmap
image/vnd.microsoft.icon
application/ico
image/ico
image/icon
text/ico
image/tiff
image/webp
audio/x-riff
```

Not sure why you see `application/x-qt-image` or even `application/x-qt-windows-mime;value="PNG"` but no usable image format.

Can you check if in your CopyQ installation you have `imageformats` **folder** containing some DLL files?

BTW, I think it should at least be possible to transform the `application/x-qt-windows-mime;value="PNG"` when available to `image/png` automatically with the following command ([here is how to add the command to CopyQ](http://copyq.readthedocs.io/en/latest/faq.html#how-to-load-shared-commands-and-share-them)):

```ini
[Command]
Automatic=true
Command="
    copyq:
    setData('image/png', input())"
Icon=\xf1c5
Input="application/x-qt-windows-mime;value=\"PNG\""
Name=Store Windows PNG
```

**Comment by iG8R** [comment]
> I don't think I can do anything about this since the clipboard (and basically Qt framework that CopyQ uses) does not seem to provide the image data.

But what happened? It used to work before, definitely in 3.x versions.
Moreover, this bug only applies to copying in browsers, any of them.
Copying from any other apps, the images are saved as images.
And also this bug with the inability to copy any images larger than a certain size. https://github.com/hluk/CopyQ/issues/2523
It seems that it's all interconnected.

Unfortunately, the `copyq clipboard \?` command does not work the same way in Windows as it does in Linux.
![image](https://github.com/hluk/CopyQ/assets/11407417/ce053184-3003-48ed-aed4-542d8d0b3cbf)

In the `imageformats` folder, I have the following files:
```
qsvg.dll
qwebp.dll
qwbmp.dll
qtiff.dll
qtga.dll
qicns.dll
qjpeg.dll
qico.dll
qgif.dll
```

> BTW, I think it should at least be possible to transform the `application/x-qt-windows-mime;value="PNG"` when available to `image/png` automatically with the following command ([here is how to add the command to CopyQ](http://copyq.readthedocs.io/en/latest/faq.html#how-to-load-shared-commands-and-share-them)):
> 
> ```ini
> [Command]
> Automatic=true
> Command="
>     copyq:
>     setData('image/png', input())"
> Icon=\xf1c5
> Input="application/x-qt-windows-mime;value=\"PNG\""
> Name=Store Windows PNG
> ```

Oh, thank you very much! And this already works in Edge and Chrome! But Firefox is overboard :(((((

**Comment by iG8R** [comment]
Found something, maybe it will somehow give you some idea on how to solve the pressing problem.
https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/clipboard/setImageData

Topics, that led to the aforementioned link:
https://github.com/Xpra-org/xpra-html5/issues/71
https://github.com/w3c/clipboard-apis/issues/44

**Comment by hluk** [comment]
> But what happened? It used to work before, definitely in 3.x versions.

Old versions of CopyQ rendered HTML items with images (and any other remote content). New versions ([since v4.0.0](https://github.com/hluk/CopyQ/commit/5746389b)) uses simpler rendering and does not make any network requests so the images show up as a dummy icon.

> Unfortunately, the `copyq clipboard \?` command does not work the same way in Windows as it does in Linux.

The console output on Windows does not work for GUI apps unfortunately (https://github.com/hluk/CopyQ/issues/349).

You can use the Action dialog (F5 shortcut) in CopyQ instead with "Store standard output" set to "text/plain" to store the command output to new item.

**Comment by hluk** [comment]
Below is a command that tries to extract the image URL from HTML and fetch the image.

A couple of downsides:
- it uses a regular expression to find the image URL in HTML which may not work in some cases
- it always stores the image with "image/png" format, this works in CopyQ even if the image format is JPEG but may fail when pasting to some apps (in some cases it would be possible to guess the format from URL itself but not always)
- it does not work with HTTP redirects

```ini
[Command]
Automatic=true
Command="
    copyq:
    const regex = /<img [^>]*src=\"([^\"]+)\"/i

    if ( dataFormats().indexOf(mimeHtml) == -1
      || dataFormats().indexOf(mimeText) != -1
      || dataFormats().indexOf('image/png') != -1)
    {
        abort();
    }

    const html = str(data(mimeHtml))

    const m = html.match(regex)
    if (!m) {
        abort();
    }

    const url = decodeURIComponent(m[1]);
    var response = networkGet(url)
    if (response.status != 200) {
        abort();
    }
    setData('image/png', response.data)"
Icon=\xf1c5
Name=Fetch Image from HTML
```

**Comment by iG8R** [comment]
[CopyQ-Image-copy-in-Firefox-Video_2023-11-03_001848.webm](https://github.com/hluk/CopyQ/assets/11407417/cd1b5d9a-5b3f-42db-8778-835b44965aa1)

**Comment by xkonglong** [comment]
Windows &  WebP  

if  OS is Windows ,  the image format is WebP.

CopyQ  can not show this image.

**Comment by MarGeb1** [comment]
Dear hluk,

I'm using CopyQ about two weeks now and I like it very much, but now I ran into the "image-gate" and found out, that this problem exists since 2022. Do you think that there will be a solution for this, because a clipboard manager that doesn't like images, is not exactly what I was looking for? 

All I can contribute is:

- I tested with Firefox, Opera and Edge on Win 10, all with the same result, no images in CopyQ, only white icons
- the problem does not exist on CopyQ for Linux (running Mint 22 Cinnamon here)
- your "Fetch Image from HTML" command didn't help and last not least
- Clipdiary 5.7 and even the 20 years old and outdated Real Clipboard can store the images.

If I can somehow assist you, just let me know.

Markus

> Below is a command that tries to extract the image URL from HTML and fetch the image.

**Comment by hluk** [comment]
> your "Fetch Image from HTML" command didn't help and last not least

@MarGeb1 What is the clipboard content and specifically value of text/html? (See the formats in File - Show Clipboard Content dialog.)

> Do you think that there will be a solution for this, because a clipboard manager that doesn't like images, is not exactly what I was looking for?

CopyQ usually works well with images. If you know about a way to get an image (without doing network requests) from a system clipboard without any known image formats, please let me know.

**Comment by MarGeb1** [comment]
That's the output for the Mint desktop picture on https://linuxmint.com/edition.php?id=316:

![Clipboard 1](https://github.com/user-attachments/assets/e3c199b1-afaf-4250-ad02-fd0b54924a89)


> @MarGeb1 What is the clipboard content and specifically value of text/html? (See the formats in File - Show Clipboard Content dialog.)

**Comment by hluk** [comment]
Not sure why it does not work for you. The regular expressing `/<img [^>]*src=\"([^\"]+)\"/i` should match the value of `src` in the HTML (and the command should fetch the image).

Is there any error in logs?

Can you try to remove the command and add it again?

Do you use the latest CopyQ version?

Interestingly, there is also `application/x-qt-image` format. Not sure why Qt framework does not also provide some `image/*` format in that case (it is usually able to convert to multiple image formats).

**Comment by MarGeb1** [comment]
That's the command I used, is it correct?

`[Command]
Automatic=true
Command="
    copyq:
    const regex = /<img [^>]*src=\"([^\"]+)\"/i

    if ( dataFormats().indexOf(mimeHtml) == -1
      || dataFormats().indexOf(mimeText) != -1
      || dataFormats().indexOf('image/png') != -1)
    {
        abort();
    }

    const html = str(data(mimeHtml))

    const m = html.match(regex)
    if (!m) {
        abort();
    }

    const url = decodeURIComponent(m[1]);
    var response = networkGet(url)
    if (response.status != 200) {
        abort();
    }
    setData('image/png', response.data)"
Icon=\xf1c5
Name=Fetch Image from HTML`

I re-added it again but with no success.

During my tests yesterday i had several

CopyQ Warning [2024-12-12 16:35:19.430] <monitorClipboard-14896>: [qt.qpa.mime] QtWarning: Retrying to obtain clipboard.
CopyQ Warning [2024-12-12 20:21:21.133] <monitorClipboard-4948>: [qt.qpa.mime] QtWarning: Retrying to obtain clipboard.

warnings, but not today.

**Comment by MarGeb1** [comment]
Here the complete warnings, sorry:

```
CopyQ Warning [2024-12-12 16:35:19.430] <monitorClipboard-14896>: [qt.qpa.mime] QtWarning: Retrying to obtain clipboard.
CopyQ Warning [2024-12-12 20:21:21.133] <monitorClipboard-4948>: [qt.qpa.mime] QtWarning: Retrying to obtain clipboard.

```

**Comment by hluk** [comment]
Can you try the following command to troubleshoot? It would show a popup with details if it cannot fetch an image or if it succeeds.

```ini
[Command]
Automatic=true
Command="
    copyq:
    const regex = /<img [^>]*src=\"([^\"]+)\"/i;

    if ( dataFormats().indexOf(mimeHtml) == -1 ) {
        popup('No HTML');
        abort();
    }

    if ( dataFormats().indexOf(mimeText) != -1 ) {
        popup('Contains text');
        abort();
    }

    if ( dataFormats().indexOf('image/png') != -1 ) {
        popup('Contains image');
        abort();
    }

    const html = str(data(mimeHtml));

    const m = html.match(regex);
    if (!m) {
        popup('No image available', html);
        abort();
    }

    const url = decodeURIComponent(m[1]);
    var response = networkGet(url);
    if (response.status != 200) {
        popup('Failed to fetch the image', `${url}\\nstatus code: ${response.status}`);
        abort();
    }

    popup('Adding an image', url);
    setData('image/png', response.data);"
Icon=
Name=Fetch Image from HTML
```

**Comment by MarGeb1** [comment]
Thank you hluk.

I copied some pictures from different sites, the result was always the same:

![Clipboard 1](https://github.com/user-attachments/assets/5bae9a37-d62c-452c-ba01-30a796707d33)

**Comment by hluk** [comment]
Hmm, that is strange. According to the previous clipboard content screenshot, there should be no `text/plain` format.

Can you try to update the command to the following? Specifically, it replaces `dataFormats().indexOf(mimeText) != -1` check with `str(data(mimeText)).trim() != ''`.

```ini
[Command]
Automatic=true
Command="
    copyq:
    const regex = /<img [^>]*src=\"([^\"]+)\"/i;

    if ( dataFormats().indexOf(mimeHtml) == -1 ) {
        popup('No HTML');
        abort();
    }

    if ( str(data(mimeText)).trim() != '' ) {
        popup('Contains text');
        abort();
    }

    if ( dataFormats().indexOf('image/png') != -1 ) {
        popup('Contains image');
        abort();
    }

    const html = str(data(mimeHtml));

    const m = html.match(regex);
    if (!m) {
        popup('No image available', html);
        abort();
    }

    const url = decodeURIComponent(m[1]);
    var response = networkGet(url);
    if (response.status != 200) {
        popup('Failed to fetch the image', `${url}\\nstatus code: ${response.status}`);
        abort();
    }

    popup('Adding an image', url);
    setData('image/png', response.data);"
Icon=
Name=Fetch Image from HTML
```

**Comment by MarGeb1** [comment]
Im really sorry but there's no difference:

![Clipboard 2](https://github.com/user-attachments/assets/8a003e36-2dd3-45ca-89d7-6fc3f7ed7259)

**Comment by hluk** [comment]
Can you please check what text (`text/plain` format) do the items contain? Select an item and press F4 to see the formats.

**Comment by MarGeb1** [comment]
text/plain is file:///C:/Users/Markus/AppData/Local/Temp/4402013300970224452.jpg
and similar entries for the other images.

**Comment by hluk** [comment]
> text/plain is file:///C:/Users/Markus/AppData/Local/Temp/4402013300970224452.jpg

Interesting, so I guess you could also use that file as image.

Anyway, here is the fixed version which fetches the remote image:

```ini
[Command]
Automatic=true
Command="
    copyq:
    const regex = /<img [^>]*src=\"([^\"]+)\"/i;

    if ( dataFormats().indexOf(mimeHtml) == -1 ) {
        abort();
    }

    const text = str(data(mimeText)).trim();
    if ( text != '' || text.startsWith('file://') ) {
        abort();
    }

    if ( dataFormats().indexOf('image/png') != -1 ) {
        abort();
    }

    const html = str(data(mimeHtml));

    const m = html.match(regex);
    if (!m) {
        abort();
    }

    const url = decodeURIComponent(m[1]);
    serverLog('Fetching image: ' + url);
    const response = networkGet(url);
    if (response.status != 200) {
        popup('Failed to fetch the image', `${url}\\nstatus code: ${response.status}`);
        abort();
    }

    setData('image/png', response.data);"
Icon=
Name=Fetch Image from HTML
```

**Comment by MarGeb1** [comment]
I'm sorry again...
If you feel bothered, I can go on with Clipdiary until I move to Linux next year. Then I will use CopyQ again because this issue does not exist there.
![Clipboard 1](https://github.com/user-attachments/assets/bf859c43-463d-485e-9250-baaeb8fa4ed2)

**Comment by hluk** [comment]
Oops, sorry, the second `if` condition in the last command was wrong, 🤦‍♂️ it should be: `text != '' && !text.startsWith('file://')`

I've updated the command in my previous comment.

**Comment by MarGeb1** [comment]
Aaaaah, looks much better now! Worked with three of my four test sites. Facebook did not, but that's not important.
Thanks for your support and have Happy Hollidays!!

![Clipboard 1](https://github.com/user-attachments/assets/1a3c88f5-de58-4560-95d2-c3687c9ca67a)

**Comment by hluk** [comment]
Oh, for Facebook, I think the command would need to go through some HTTP redirects or even log into an account.

It would be much better to use the image path if it is in the `text/plain` format. And I just remembered, there is even a command for this: [Preview Image Files](https://github.com/hluk/copyq-commands/tree/master/Display#preview-image-files)

**Comment by MarGeb1** [comment]
That does not work 🤪. But as said, I really don't need Facebook.

![Clipboard 1](https://github.com/user-attachments/assets/8d4d445a-ad3b-4371-91aa-e1b1197fd842)

**Comment by esauvisky** [comment]
Found a solution that doesn't require fetching:
```javascript
global.hasImageFormat = function (formats) {
    for (const format of formats.values()) {
        if (format.startsWith('image/'))
            return true;
    }
    return false;
};

const imageMime = 'image/png';
const mimeText = 'text/plain';
const onClipboardChanged_ = onClipboardChanged;
onClipboardChanged = function () {
    if (isClipboard()) {
        const formats = dataFormats();
    
        // Fix for Chrome image issue
        const imageData = clipboard(imageMime);
        if (imageData.length > 0 || hasImageFormat(formats)) {
            removeData(mimeText)
            setData(imageMime, imageData);
        }
    }
    onClipboardChanged_();
}
```

**Comment by Mattwmaster58** [comment]
my version which can handle remote, local, and data URLs:
```js
copyq:
const regex = /<img [^>]*src="([^"]+)"/i;

if ( dataFormats().indexOf(mimeHtml) == -1 ) {
    serverLog("no HTML mimetype available");
    abort();
}

const html = str(data(mimeHtml));
const imgSrc = html.match(regex)
if (!imgSrc) {
    serverLog("img not found within html", html);
    abort();
}

const text = str(data(mimeText)).trim();
if ( text != '' && !text.match(/^(data:image\/\w+;base64|file:\/\/|http)/i) ) {
    serverLog("text complement didn't match", text);
    abort();
}

const url = decodeURIComponent(imgSrc[1]);
serverLog('Fetching image: ' + url);
const response = networkGet(url);
if (response.status != 200) {
    serverLog('Failed to fetch the image for preview', `${url}\nstatus code: ${response.status}`);
    abort();
}

// to be closer to the truth we would have to consider the response header and guess in the case of missing response
// applications handle this well enough, even when the specific image mime type does not match
setData('image/png', response.data);
```

**Comment by esauvisky** [comment]
Found a cleaner solution:

```ini
[Command]
Command="
    const mimeImage = 'image/png';
    const _onClipboardChanged = onClipboardChanged;
    onClipboardChanged = function () {
        if (hasClipboardFormat(mimeImage)) {
            _onClipboardChanged();
            imageData = clipboard(mimeImage);
            setData(mimeImage, imageData);
            saveData();
            return;
        }
        _onClipboardChanged();
    }"
Icon=
IsScript=true
Name=Fix Image Copy
```

@hluk Any thoughts on this issue? It appears the image data can only be obtained by calling `clipboard()` inside `onClipboardChanged()`, leading to lots of weird race conditions if you have complex logic within that hook like I do. The solution above isn't perfect either, sometimes it overwrites the data of the last item when calling `setData`.

This issue also isn't limited to browsers, it occurs in GNOME with most GTK applications, for example, such as `eog` (the default image viewer).

**Comment by hluk** [comment]
> It appears the image data can only be obtained by calling `clipboard()` inside `onClipboardChanged()`, leading to lots of weird race conditions if you have complex logic within that hook like I do.

So the image is available in the system clipboard? Maybe it "appears" later? Are there any errors in logs?

To avoid adding a different clipboard data to the currently captured data for `onClipboardChanged`, you would need to check to avoid the data race:

```js
const mimeImage = 'image/png';
const _onClipboardChanged = onClipboardChanged;
onClipboardChanged = function() {
    const imageData = clipboard(mimeImage);
    if (imageData.length > 0
        && str(data(mimeText)) === str(clipboard(mimeText))
        && str(data(mimeHtml)) === str(clipboard(mimeHtml))
    ) {
        setData(mimeImage, imageData);
    }
    _onClipboardChanged();
}
```

**Comment by Chealer** [comment]
@Searcher1231: This report is hard to follow. Please observe [that checklist](https://blog.michael.gr/post/2014-11-bug-reporting-checklist/), in particular sections 8 to 13, and indicate whether the "Show simple items" preference in _History_ is enabled.

**Comment by NA0341** [comment]
Gosh, this seems to be a never ending issue. 😆
Maybe just don't use a clipboard manager to store images you want to preserve from web.

I mean it's definitely helpful for image editing when you copy and cut image selections and want to paste them later then it's really great but for long term storage I don't think it's a solution in any way really ~ because yeah you are storing that in databases along with text and lots of other stuff and copying lots of images into a database makes it pretty big pretty quick so might not be the best idea in the first place.

I phased out image storing in CopyQ a few years ago.
Now with just text my total is currently at >290MB in 77 tabs (exported .cpq file).
